ansible_base/authentication/authenticator_plugins/radius.py

```diff
@@ -0,0 +1,323 @@
+import logging
+from io import StringIO
+from types import SimpleNamespace
+
+from django.contrib.auth import get_user_model
+from django.contrib.auth.models import Group
+from django.utils.translation import gettext_lazy as _
+from pyrad.client import Client, Timeout
+from pyrad.dictionary import Dictionary
+from pyrad.packet import AccessAccept, AccessReject, AccessRequest
+
+from ansible_base.authentication.authenticator_plugins.base import AbstractAuthenticatorPlugin, BaseAuthenticatorConfiguration
+from ansible_base.authentication.utils.claims import get_or_create_authenticator_user
+from ansible_base.lib.serializers.fields import CharField, IntegerField
+
+logger = logging.getLogger(__name__)
+
+
+User = get_user_model()
+
+
+class RADIUSConfiguration(BaseAuthenticatorConfiguration):
+
+    HOST = CharField(
+        label=_("RADIUS Server"),
+        help_text="Hostname of RADIUS server.",
+        ui_field_label=_('Hostname of RADIUS Server'),
+        required=True,
+    )
+
+    PORT = IntegerField(
+        min_value=1,
+        max_value=65535,
+        default=1812,
+        label=_("RADIUS Port"),
+        help_text=_("Port number of RADIUS server."),
+        ui_field_label=_("Port number of RADIUS Server"),
+    )
+
+    SECRET = CharField(
+        label=_("RADIUS Secret"),
+        help_text=_("Shared secret for authenticating to RADIUS server."),
+        ui_field_label=_("Shared secret for authenticating to RADIUS server."),
+        required=True,
+    )
+
+
+class AuthenticatorPlugin(AbstractAuthenticatorPlugin):
+    configuration_class = RADIUSConfiguration
+    type = "RADIUS"
+    category = "password"
+    configuration_encrypted_fields = ["SECRET"]
+
+    def __init__(self, database_instance=None, *args, **kwargs):
+        super().__init__(database_instance, *args, **kwargs)
+        self.set_logger(logger)
+
+    def authenticate(self, request, username, password, **kwargs):
+        settings = SimpleNamespace(
+            RADIUS_SERVER=self.settings["HOST"],
+            RADIUS_PORT=self.settings["PORT"],
+            RADIUS_SECRET=self.settings["SECRET"],
+        )
+        backend = RADIUSBackend(settings)
+        user = backend.authenticate(request, username, password)
+        if user is None:
+            return None
+
+        get_or_create_authenticator_user(
+            user_id=username,
+            user_details={"username": username},
+            authenticator=self.database_instance,
+            extra_data={},
+        )
+        return user
+
+
+# The code below is based on the source code of the django-radius library (https://github.com/robgolding/django-radius).
```

Comment:
We already forked django-radius and use it in AWX here https://github.com/ansible/django-radius/blob/develop/radiusauth/backends/radius.py. Would it make sense to just modify that and use a git require? We could also do a PR to the actual django-radius as well.

Reply:
Longer term we also want to split out authenticators into their own repos and make a "pluggable" system like collections in Ansible. So that community members can contribute to individual plugins/write new ones.